### ISSUE TITLE
Clarify usage/support of Expect header

### ISSUE THREAD

**Issue by acoburn** [fcrepo/fcrepo-specification#87]

**Comment by ajs6f** [comment]
Since you have looked at the code, how far do you think `httpd` is from support?

**Comment by ruebot** [comment]
@awoods @zimeon @barmintor

**Comment by acoburn** [comment]
I seriously doubt that httpd will ever support expectation-extension. And it's not because supporting it is hard.

The whole expectation-extension mechanism is part of rfc 2616, but it has been completely removed from the newer http spec.

Let me reiterate that relying on this is a bad idea.

**Comment by acoburn** [comment]
From RFC 7231:
      Note: The Expect header field was added after the original
      publication of HTTP/1.1 [RFC2068] as both the means to request an
      interim 100 (Continue) response and the general mechanism for
      indicating must-understand extensions.  However, the extension
      mechanism has not been used by clients and the must-understand
      requirements have not been implemented by many servers, rendering
      the extension mechanism useless.  This specification has removed
      the extension mechanism in order to simplify the definition and
      processing of 100-continue.

**Comment by ruebot** [comment]
@acoburn do you have any recommendations for option 1, another header?

**Comment by acoburn** [comment]
@ruebot no I do not. I have not been involved with the fixity section. personally, I have no use for the fixity section-- it's better handled out-of-band, IMO

**Comment by ruebot** [comment]
@acoburn so, leave 5.1 and 5.2, but remove 5.3? If so, I support that.

**Comment by acoburn** [comment]
@ruebot yes remove 5.3, but keep 5.1 and 5.2

**Comment by zimeon** [comment]
Well, the comment in RFC 7231 does seem to completely kill the current spec for persistence fixity, I note that also means [3.5.2.1 Expect: 202-digest](http://fedora.info/spec/#httpGETLDPNR-fixity-expectation) (and sub-section 3.5.2.1.1) must go or be changed too.

On what basis do we decide whether persistence fixity should be entirely removed or not?

**Comment by ruebot** [comment]
@zimeon I've been a supporter of persistence fixity going for a while. I think it is better done elsewhere in the stack.

**Comment by acoburn** [comment]
I would like to hear from @barmintor, but I don't see how persistence fixity can remain as long as it uses the Expect header.

**Comment by zimeon** [comment]
I agree that it can't remain in current form (`Expect` approach dead) but I feel an uneasiness suddenly switching from working out how to make it work to ditching it, a significant change in functionality.

**Comment by acoburn** [comment]
One option, like with batch ops, is to move this portion to an external non-core spec where it can be developed independently

**Comment by barmintor** [comment]
@acoburn is right, we need to move persistence fixity out given that note. But I think this *also* destabilizes `Want-Digest`- do we need to clarify that, as it applies to the message identified by the URI, that it is a current content digest?

**Comment by ruebot** [comment]
PR: https://github.com/fcrepo/fcrepo-specification/pull/88

**Comment by birkland** [comment]
This issue is moving a little bit fast - I don't understand the intentions related to persistence fixity.  Atomic batch ops were moved out of the specification due to concerns related to underling principles and semantics (rather than API details).  It seems like the problem in this ticket is related API details rather than a concern over whether the spec should contain persistence fixity in the first place.  Removing persistence fixity in its entirety feels like a non-sequitur, at least based on the comments of this ticket so far. 

 I have the same question as @zimeon: On what basis do we decide whether persistence fixity should be entirely removed or not?  @ruebot @acoburn : Can you elaborate more on "better done elsewhere in the stack" or "better done out-of-band"?

**Comment by acoburn** [comment]
@birkland there are persistence layers that already provide end-to-end data integrity guarantees. ZFS is one such filesystem that does this; there are others, too, but that's the one I'm most familiar with. Why should Fedora do the same thing? It is redundant and inefficient. If users need these guarantees (we certainly want this at Amherst), then they should choose systems that already have this built in rather than relying on a mechanism at a higher level in the stack that is considerably less efficient.

IMO, persistence fixity is completely separate from transport fixity, and the fact that we are wedging persistence fixity into an HTTP interaction (which is fundamentally about transport) makes very little sense to me.

**Comment by ruebot** [comment]
@birkland exactly what @acoburn said.

**Comment by awoods** [comment]
Although I agree that there are backend storage systems that take responsibility for ensuring data integrity, there have been (albeit, undocumented) user requests for the Fedora repository capability of reporting back on such data integrity. Doron Shalvi comes to mind.

It may be that we determine that the functionality should be removed, but as @birkland and @zimeon note, the basis for removing the functionality should be grounded in Fedora repository capability expectations,
not the fact that a particular header extension is invalid.

I suggest we explore @acoburn's first option:
> 1. Choose another header.

@barmintor : can you elaborate on your comment above:
> But I think this also destabilizes `Want-Digest`

**Comment by acoburn** [comment]
@awoods:

> there have been (albeit, undocumented) user requests for the Fedora repository capability of reporting back on such data integrity

Is this a legitimate reason for including a feature in the Fedora specification? I thought the bar was a bit higher.

**Comment by acoburn** [comment]
End-to-end data integrity is a perfectly legitimate concern.

That said, the basic premise of the specification, as written, is short-sighted. It is predicated on the notion of a file-based persistence strategy; that is, a strategy where data integrity can be efficiently computed for a given file-based resource.

This is a false notion.

While it is true that a one-file-per-binary-resource persistence strategy is certainly an easy one to implement, it is by no means the only way to construct a persistence strategy. Nor it is necessarily a good way to ensure data integrity.

For example: what if the underlying data layer relies on merkle trees? Or a block chain? At that point, how does _persistence_ integrity differ in a meaningful sense from _transport_ integrity?

In my opinion the notion of ensuring persistence-level data integrity over a RESTful (HTTP) interface is fraught, and it will be a continual problem for the Fedora specification.

I now have two Fedora implementations in progress and while they will both support transport-level data integrity checks (via `Want-Digest`), neither will offer "persistence fixity support".

**Comment by awoods** [comment]
Thanks, @acoburn. The premise of the fixity specification is that a binary resource has a given state at a given time, versus assuming it is file-based. 

It would be helpful if you would elaborate on how you envision persisting binary resources such that calculating fixity of their state is not possible.

**Comment by acoburn** [comment]
@awoods perhaps you misunderstand me. I am not suggesting that computing fixity for a resource is not possible. Rather it is neither efficient nor is it in scope for an HTTP server endpoint. Furthermore, there is already a mechanism in the specification for computing the fixity for a resource: `Want-Digest`.

**Comment by awoods** [comment]
In order to clearly tease out the two threads of this discussion, @acoburn, would you be willing to close this issue with the creation of two new issues:
1. `Expect: 202-digest` is not supported per RFC 7231 - need alternate approach
2. Remove persistence-fixity - inefficient and out of scope

The trouble with this current thread is that it uses the fact that `Expect: 202-digest` is not supported as the basis of the argument for removing the functionality. Finding a supported approach for persistence-fixity is a separate topic from removing the capability.

**Comment by barmintor** [comment]
@awoods all I meant by destabilizing was the question that follows it in that comment re: clarifying what is expected.

**Comment by birkland** [comment]
To me, it's only worth fixing the API details of persistence fixity if the notion of persistence-fixity-as-a-service makes sense - that's what is fundamentally being questioned here.  So the meat of this issue belies its title and description, and I like splitting it into two tickets.

The underlying premise behind persistence fixity in the spec (as I understand it) is:
The repository is uniquely positioned to compute or otherwise engage storage implementations to verify persistence fixity of binary resources, and presents a simple API for doing so.  By being close to the storage, or able to interact with storage via native APIs, the repository can provide a simple and efficient abstraction for fixity verification.  This abstraction is useful to clients that enact fixity related policies, e.g. periodic validation, auditing, etc.

Whether or not we develop an alternate approach depends on whether the above premise holds, in my opinion.

**Comment by ajs6f** [comment]
This is another example of the tension between 

1. API as minimal suite of functions necessary and sufficient to "do Fedora" and
2. API as very small suite of functions that are very likely to be useful to someone "doing Fedora" in the sense that someone already "doing Fedora" has provided a use case for each.

It was never clear, from the beginning of this effort, which goal was being pursued, and the first only makes sense if you start from some pre-agreed notion of what "doing Fedora" is.

**Comment by barmintor** [comment]
For reference, the last round of questions about fixity's inclusion in the spec:
https://wiki.duraspace.org/display/FF/2016-12-08+-+Fedora+Tech+Meeting
https://groups.google.com/d/msg/fedora-tech/HMpD9n4273U/hOWX5WQ2CgAJ
https://wiki.duraspace.org/display/FF/2017-01-09+Fixity+Service+Special+Topic+meeting
https://docs.google.com/document/d/1Nm-LFyUhCl-et4UsdePuMJ_Zu23Mot4FiEpEMke_JH4/edit

**Comment by barmintor** [comment]
@awoods the `Want-Digest` clarification is that the digest needs to be currently calculated if the goal of providing digest verification without transfer is to be met.

**Comment by acoburn** [comment]
@awoods: I will close this issue once #88 is merged.

**Comment by awoods** [comment]
@acoburn : #88 is a PR that removes persistence-fixity from the spec. The suggestion I am making is that we create two new issues to discuss:
1. whether persistence-fixity should be removed from the spec, and if not
2. what alternate approach could be used in place of `Expect: 202-digest`

At a minimum, we should have a new issue for the discussion of whether persistence-fixity should be removed. This current issue is too intertwined with `Expect: 202-digest` to be effective in that regard.

**Comment by acoburn** [comment]
@awoods I am going to hold firm on this. The various sections of the Fedora specifications were drafted months ago; the current language is incompatible with HTTP/1.1; therefore it needs to be removed.

Once that language is removed, please feel free to open a new issue to discuss whether persistence-fixity belongs in the spec. If it is decided to add that language back into the spec, then do so, but I am strongly in favor of dropping that section.

**Comment by ajs6f** [comment]
@awoods You are confusing "remove this language" and "remove fixity from the spec". @acoburn is arguing for the former (which will have to take place in any event at all), not the latter. The latter is the separate discussion that you want and against which no one is arguing.

**Comment by awoods** [comment]
@acoburn, thanks. I understand your position. Before entertaining the idea of removing a capability from the spec... we should discuss whether it should be removed.

The fact that the current draft recommends a non-supported approach per RFC 7231 is not basis for removing the capability... but it is grounds for correcting the approach.

**Comment by acoburn** [comment]
@awoods that is fair, but in the meantime, people are trying to implement the spec as written. You cannot have language in the spec that is so clearly wrong.

**Comment by awoods** [comment]
@acoburn : Agreed, we should get these two issues resolved as soon as possible. My observation is that this current issue is too long and intertwined to expect engagement from others... hence the suggestion to break it out into the two, clear issues at hand:
1. whether persistence-fixity should be removed from the spec, and if not
2. what alternate approach could be used in place of `Expect: 202-digest`

**Comment by acoburn** [comment]
@awoods: Then merge #88 and open two new issues. Or, if one other person gives a 👍  to #88, I'll merge it. Then we can discuss the path forward.

**Comment by awoods** [comment]
@acoburn : It seems like bad form to suggest merging a PR that is "in debate".

**Comment by acoburn** [comment]
@awoods: as @ajs6f has already pointed out

> You are confusing "remove this language" and "remove fixity from the spec". @acoburn is arguing for the former (which will have to take place in any event at all), not the latter. The latter is the separate discussion that you want and against which no one is arguing.
